### ISSUE TITLE
🧠 Workstream B.1: ML Pipeline Optimization (torch.compile, no ONNX)

### ISSUE THREAD

**Issue by Shukik85** [Shukik85/hydraulic-diagnostic-saas#31]
## Описание
Оптимизация ML inference pipeline с использованием PyTorch torch.compile для ускорения вывода GNN моделей в production. Реализация batch async API, метрики latency через Prometheus, benchmark тесты.

## Цели
- [x] Применить torch.compile для ускорения inference
- [x] Реализовать batch async inference endpoint
- [x] Добавить Prometheus latency metrics
- [x] Провести benchmark: baseline vs torch.compile
- [x] Задокументировать результаты оптимизации

### 1. torch.compile Optimization

```python
# services/ml_service/inference.py
import torch
import torch.nn as nn
from typing import List, Dict
import time

class GNNInferenceService:
    def __init__(self, model_path: str, device: str = "cuda"):
        self.device = torch.device(device)
        
        # Load model
        self.model = self.load_model(model_path)
        self.model.eval()
        self.model.to(self.device)
        
        # Apply torch.compile optimization
        # mode: "default" | "reduce-overhead" | "max-autotune"
        self.model = torch.compile(
            self.model,
            mode="max-autotune",  # Best for production throughput
            fullgraph=True
        )
        
        # Warmup для компиляции
        self._warmup()
    
    def _warmup(self, num_iterations: int = 10):
        """Warmup для torch.compile - первые вызовы компилируют граф"""
        dummy_input = self._create_dummy_input()
        
        with torch.inference_mode():
            for _ in range(num_iterations):
                _ = self.model(dummy_input)
        
        torch.cuda.synchronize()  # Дождаться завершения компиляции
    
    @torch.inference_mode()
    def predict_batch(self, batch: List[Dict]) -> List[Dict]:
        """Batch inference с torch.compile optimization"""
        # Preprocessing
        x = self._preprocess_batch(batch)
        x = x.to(self.device)
        
        # Inference
        start = time.perf_counter()
        outputs = self.model(x)
        torch.cuda.synchronize()  # Важно для точного замера
        latency = time.perf_counter() - start
        
        # Postprocessing
        results = self._postprocess_outputs(outputs)
        
        return results, latency
```

### 2. FastAPI Batch Async Endpoint

```python
# api/inference.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel, Field
from typing import List, Dict, Optional
from prometheus_client import Histogram, Counter
import asyncio

app = FastAPI()

# Prometheus metrics
inference_latency = Histogram(
    'ml_inference_latency_seconds',
    'ML model inference latency',
    buckets=[0.01, 0.05, 0.1, 0.5, 1.0, 2.0, 5.0]
)

inference_requests = Counter(
    'ml_inference_requests_total',
    'Total inference requests',
    ['batch_size', 'status']
)

batch_size_histogram = Histogram(
    'ml_inference_batch_size',
    'Batch size distribution',
    buckets=[1, 8, 16, 32, 64, 128, 256]
)

class InferenceRequest(BaseModel):
    telemetry_data: List[Dict]
    equipment_id: str
    session_id: Optional[str] = None
    
    class Config:
        json_schema_extra = {
            "example": {
                "telemetry_data": [
                    {"pressure": 120, "temp": 45, "flow": 85},
                    {"pressure": 118, "temp": 46, "flow": 87}
                ],
                "equipment_id": "hyd-001",
                "session_id": "diag_20251114_143501"
            }
        }

class InferenceResponse(BaseModel):
    predictions: List[Dict]
    anomaly_score: float = Field(..., ge=0.0, le=1.0)
    severity: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    model_version: str
    inference_time_ms: float
    batch_size: int

@app.post("/v1/inference/batch", response_model=InferenceResponse)
async def batch_inference(request: InferenceRequest):
    """
    Batch async inference endpoint with torch.compile optimization
    """
    batch_size = len(request.telemetry_data)
    batch_size_histogram.observe(batch_size)
    
    try:
        # Run inference in thread pool (torch не полностью async)
        results, latency = await asyncio.to_thread(
            app.state.inference_service.predict_batch,
            request.telemetry_data
        )
        
        # Record metrics
        inference_latency.observe(latency)
        inference_requests.labels(
            batch_size=str(batch_size),
            status='success'
        ).inc()
        
        return InferenceResponse(
            predictions=results['predictions'],
            anomaly_score=results['anomaly_score'],
            severity=results['severity'],
            confidence=results['confidence'],
            model_version=app.state.model_version,
            inference_time_ms=latency * 1000,
            batch_size=batch_size
        )
        
    except Exception as e:
        inference_requests.labels(
            batch_size=str(batch_size),
            status='error'
        ).inc()
        raise HTTPException(status_code=500, detail=str(e))

# Health check для K8s
@app.get("/health/startup")
async def startup_health():
    """Startup probe - model loaded and compiled"""
    if not hasattr(app.state, 'inference_service'):
        return {"status": "initializing"}, 503
    return {"status": "ready"}

@app.get("/health/live")
async def liveness():
    return {"status": "alive"}

@app.get("/health/ready")
async def readiness():
    """Readiness probe - GPU available, model loaded"""
    try:
        if not torch.cuda.is_available():
            return {"status": "no_gpu"}, 503
        
        # Quick inference test
        test_result = await asyncio.to_thread(
            app.state.inference_service._warmup,
            num_iterations=1
        )
        return {"status": "ready"}
    except Exception as e:
        return {"status": "not_ready", "error": str(e)}, 503
```

### 3. Benchmark Script

```python
# benchmarks/inference_benchmark.py
import torch
import time
import numpy as np
from typing import List, Tuple
import pandas as pd
import matplotlib.pyplot as plt

class InferenceBenchmark:
    def __init__(self, model_path: str):
        self.model = self.load_model(model_path)
        self.model.eval()
        self.model.cuda()
        
        # Baseline model (no optimization)
        self.baseline_model = self.model
        
        # torch.compile optimized
        self.compiled_model = torch.compile(
            self.model,
            mode="max-autotune"
        )
        
        # Warmup
        self._warmup(self.baseline_model)
        self._warmup(self.compiled_model)
    
    def benchmark(
        self,
        batch_sizes: List[int] = [1, 8, 16, 32, 64, 128],
        num_iterations: int = 100
    ) -> pd.DataFrame:
        """
        Benchmark baseline vs torch.compile
        """
        results = []
        
        for batch_size in batch_sizes:
            print(f"\\nBenchmarking batch_size={batch_size}")
            
            # Create dummy input
            dummy_input = torch.randn(batch_size, 128, device='cuda')
            
            # Baseline
            baseline_latencies = []
            for _ in range(num_iterations):
                start = time.perf_counter()
                with torch.inference_mode():
                    _ = self.baseline_model(dummy_input)
                torch.cuda.synchronize()
                baseline_latencies.append(time.perf_counter() - start)
            
            # torch.compile
            compiled_latencies = []
            for _ in range(num_iterations):
                start = time.perf_counter()
                with torch.inference_mode():
                    _ = self.compiled_model(dummy_input)
                torch.cuda.synchronize()
                compiled_latencies.append(time.perf_counter() - start)
            
            # Statistics
            baseline_mean = np.mean(baseline_latencies)
            compiled_mean = np.mean(compiled_latencies)
            speedup = baseline_mean / compiled_mean
            
            results.append({
                'batch_size': batch_size,
                'baseline_mean_ms': baseline_mean * 1000,
                'compiled_mean_ms': compiled_mean * 1000,
                'speedup': speedup,
                'baseline_p50_ms': np.percentile(baseline_latencies, 50) * 1000,
                'compiled_p50_ms': np.percentile(compiled_latencies, 50) * 1000,
                'baseline_p99_ms': np.percentile(baseline_latencies, 99) * 1000,
                'compiled_p99_ms': np.percentile(compiled_latencies, 99) * 1000,
            })
            
            print(f"  Baseline: {baseline_mean*1000:.2f}ms")
            print(f"  Compiled: {compiled_mean*1000:.2f}ms")
            print(f"  Speedup: {speedup:.2f}x")
        
        return pd.DataFrame(results)
    
    def plot_results(self, df: pd.DataFrame, save_path: str = "benchmark_results.png"):
        """Plot benchmark results"""
        fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5))
        
        # Latency comparison
        ax1.plot(df['batch_size'], df['baseline_mean_ms'], 'o-', label='Baseline')
        ax1.plot(df['batch_size'], df['compiled_mean_ms'], 's-', label='torch.compile')
        ax1.set_xlabel('Batch Size')
        ax1.set_ylabel('Latency (ms)')
        ax1.set_title('Inference Latency: Baseline vs torch.compile')
        ax1.legend()
        ax1.grid(True)
        
        # Speedup
        ax2.bar(df['batch_size'], df['speedup'])
        ax2.axhline(y=1.0, color='r', linestyle='--', label='No speedup')
        ax2.set_xlabel('Batch Size')
        ax2.set_ylabel('Speedup (x)')
        ax2.set_title('torch.compile Speedup')
        ax2.legend()
        ax2.grid(True)
        
        plt.tight_layout()
        plt.savefig(save_path, dpi=150)
        print(f"\\nResults saved to {save_path}")

if __name__ == "__main__":
    benchmark = InferenceBenchmark("models/gnn_model.pt")
    results = benchmark.benchmark()
    
    print("\\n=== Benchmark Results ===")
    print(results.to_string(index=False))
    
    benchmark.plot_results(results)
    results.to_csv("benchmark_results.csv", index=False)
```

### 4. Prometheus Metrics Export

```python
# api/metrics.py
from prometheus_client import make_asgi_app

# Mount Prometheus metrics endpoint
metrics_app = make_asgi_app()
app.mount("/metrics", metrics_app)
```

### 5. Load Test Script

```python
# tests/load_test.py
import asyncio
import aiohttp
import time
import numpy as np
from typing import List

async def run_load_test(
    url: str = "http://localhost:8001/v1/inference/batch",
    batch_size: int = 32,
    num_requests: int = 1000,
    concurrent_requests: int = 10
):
    """
    Load test для batch inference endpoint
    """
    latencies = []
    errors = 0
    
    async def send_request(session, request_id):
        # Generate dummy telemetry
        telemetry = [
            {
                "pressure": 100 + np.random.randn() * 10,
                "temperature": 45 + np.random.randn() * 5,
                "flow_rate": 85 + np.random.randn() * 8
            }
            for _ in range(batch_size)
        ]
        
        payload = {
            "telemetry_data": telemetry,
            "equipment_id": f"hyd-{request_id % 100}",
            "session_id": f"test_{request_id}"
        }
        
        try:
            start = time.perf_counter()
            async with session.post(url, json=payload) as resp:
                result = await resp.json()
                latency = time.perf_counter() - start
                latencies.append(latency)
                
                if resp.status != 200:
                    return False
                return True
        except Exception as e:
            print(f"Request {request_id} failed: {e}")
            return False
    
    # Run load test
    async with aiohttp.ClientSession() as session:
        start_time = time.time()
        
        tasks = []
        for i in range(num_requests):
            task = send_request(session, i)
            tasks.append(task)
            
            # Control concurrency
            if len(tasks) >= concurrent_requests:
                results = await asyncio.gather(*tasks)
                errors += sum(1 for r in results if not r)
                tasks = []
        
        # Remaining tasks
        if tasks:
            results = await asyncio.gather(*tasks)
            errors += sum(1 for r in results if not r)
        
        total_time = time.time() - start_time
    
    # Statistics
    success_rate = (num_requests - errors) / num_requests * 100
    throughput = num_requests / total_time
    
    latencies = np.array(latencies) * 1000  # to ms
    
    print(f"\\n=== Load Test Results ===")
    print(f"Total requests: {num_requests}")
    print(f"Concurrent: {concurrent_requests}")
    print(f"Batch size: {batch_size}")
    print(f"Success rate: {success_rate:.2f}%")
    print(f"Throughput: {throughput:.2f} req/s")
    print(f"Latency p50: {np.percentile(latencies, 50):.2f}ms")
    print(f"Latency p95: {np.percentile(latencies, 95):.2f}ms")
    print(f"Latency p99: {np.percentile(latencies, 99):.2f}ms")
    
    assert success_rate > 99, f"Success rate too low: {success_rate}%"
    assert np.percentile(latencies, 99) < 500, f"P99 latency too high"

if __name__ == "__main__":
    asyncio.run(run_load_test())
```

## Критерии приемки
- ✅ torch.compile применён и показывает speedup >1.5x
- ✅ Batch async endpoint работает с FastAPI
- ✅ Prometheus metrics экспортируются (/metrics)
- ✅ Benchmark результаты задокументированы (CSV + график)
- ✅ Load test проходит с success rate >99%
- ✅ P99 latency <500ms для batch_size=32

## Ссылки
- [PyTorch torch.compile](https://pytorch.org/docs/stable/torch.compiler.html)
- [FastAPI Background Tasks](https://fastapi.tiangolo.com/tutorial/background-tasks/)
- [Prometheus Python Client](https://github.com/prometheus/client_python)

## Зависимости
- Issue #30 (K8s deployment)
- GPU-enabled K8s node
- TimescaleDB для логирования (optional)

## Оценка
⏱️ **~3 часа**

**Labels**: `prod-readiness`, `ml`, `performance`, `workstream-b`
**Parent Issue**: #27
